airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/PostStateAsyncStateManager.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination_async.state;
+
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.v0.AirbyteStreamState;
+import io.airbyte.protocol.models.v0.StreamDescriptor;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+// todo (cgardens) - hook it up to the memory manager.
+public class PostStateAsyncStateManager {
+
+  private final Map<StreamDescriptor, Queue<Long>> streamToStateId;
+  private final Map<StreamDescriptor, Long> streamToOpenState;
+  private final Map<Long, AirbyteMessage> stateIdToState;
+  private final Map<Long, AtomicLong> stateIdToCounter;
+
+  private final Set<Long> aliasIds;
+  private final long aliasedId;
+
+  private long lastStateId;
+
+  private final AirbyteStateType stateType;
+
+  public PostStateAsyncStateManager(final AirbyteStateType stateType,
+                                    final ConcurrentMap<StreamDescriptor, Long> streamToStateId,
+                                    final ConcurrentMap<Long, AtomicLong> stateIdToCounter,
+                                    final long lastStateId,
+                                    final AirbyteMessage initialState) {
+    // same regardless of state type.
+    this.stateType = stateType;
+    this.lastStateId = lastStateId;
+    this.streamToStateId = new ConcurrentHashMap<>();
+    this.streamToOpenState = new ConcurrentHashMap<>();
+    this.stateIdToState = new ConcurrentHashMap<>();
+    this.stateIdToCounter = new ConcurrentHashMap<>();
+
+    if (stateType == AirbyteStateType.STREAM) {
+      // not relevant in stream.
+      aliasedId = 0L;
+      aliasIds = new HashSet<>();
+
+      // for each stream load it into maps.
+      for (final Entry<StreamDescriptor, Long> e : streamToStateId.entrySet()) {
+        final StreamDescriptor streamDescriptor = e.getKey();
+        final long stateId = e.getValue();
+
+        // copied from registerNewStreamDescriptor
+        this.streamToStateId.put(streamDescriptor, new LinkedList<>());
+        // copied from registerNewStateId
+        this.streamToStateId.get(streamDescriptor).add(stateId);
+        this.streamToOpenState.put(streamDescriptor, stateId);
+        this.stateIdToCounter.put(stateId, new AtomicLong());
+      }
+
+      // set up initial state.
+      closeState(initialState);
+    } else {
+      // alias old stream-level state ids to single global state id
+      aliasIds = new HashSet<>(streamToStateId.values());
+      aliasedId = this.lastStateId++;
```

Comment:
because in pre state mode we had state ids PER stream, but it turned out we were global, we need to have an alias from the old ids to the single state id.